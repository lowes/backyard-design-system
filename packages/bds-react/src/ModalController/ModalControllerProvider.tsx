import * as React from 'react'

import type { ModalControllerProps } from './ModalController'

const ModalControllerContext = React.createContext<ModalControllerContextValues>(null)

const ModalControllerConsumer = ModalControllerContext.Consumer

/**
 * `ModalControllerProvider` Backyard React
 *
 * @param props - Modal props
 */
const ModalControllerProvider = ({
    children,
    override,
    context: contextProp,
    ...props
}: ModalControllerProviderProps): React.ReactElement<ModalControllerProviderProps> => {
    // Memoize `ModalControllerContextValues`
    const context = React.useMemo<ModalControllerContextValues>(
        () => ({
            ...contextProp,
            ...override,
            ...props,
        }),
        [override, props, contextProp],
    )

    return (
        <ModalControllerContext.Provider value={context}>
            {children}
        </ModalControllerContext.Provider>
    )
}

interface ModalControllerContextValues extends ModalControllerProviderProps, Record<string, any> {
    /**
     * Props to override children with
     */
    override?: Record<string, any>
}

interface ModalControllerProviderProps extends Omit<ModalControllerProps, 'modal'> {
    /**
     * Props to override children with
     */
    override?: Record<string, any>
}

export { ModalControllerContext, ModalControllerConsumer, ModalControllerProvider }

export type { ModalControllerContextValues, ModalControllerProviderProps }

export default ModalControllerProvider
