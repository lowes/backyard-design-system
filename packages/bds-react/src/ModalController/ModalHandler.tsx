import * as React from 'react'
import ReactDOM from 'react-dom'
import classNames from 'classnames'

import ownerDocument from '../utils/functions/ownerDocument'
import useForkRef from '../utils/hooks/useForkRef'
import useEventCallback from '../utils/hooks/useEventCallback'
import type { BackyardBaseProps } from '../utils/typings/BackyardProps'
import Portal from '../Portal'
import type { PortalRef } from '../Portal'
import TrapFocus from '../TrapFocus'
import Overlay from '../Overlay'
import type { OverlayProps } from '../Overlay'

import ModalManager, { ariaHidden } from './ModalManager'
import ModalHandlerWrapper from './styles/ModalHandlerWrapper'
import type { BDSForwardRef } from '../utils/typings/BDSComponentProps'

/**
 * Gets container node from DOM if needed
 *
 * @param {func|node} container
 */
function getContainer(container) {
    // eslint-disable-next-line no-param-reassign
    container = typeof container === 'function' ? container() : container

    // eslint-disable-next-line react/no-find-dom-node
    return ReactDOM.findDOMNode(container)
}

// A modal manager used to track and manage the state of open Modals.
// Modals don't open on the server so this won't conflict with concurrent requests.
const defaultManager = new ModalManager()

/**
 * Backyard React Modal Handler
 *
 * This component is used to extend the higher ordered component `ModalController`
 * Abstracts some of the complexity for future extending, if needed
 *
 * Heavily influenced by Material UI's Modal Component
 * @see https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Modal/Modal.js
 */
const ModalHandler: BDSForwardRef<ModalHandlerProps> = React.forwardRef<
    ModalHandlerRef,
    ModalHandlerProps
>(function ModalHandler(
    {
        OverlayComponent = Overlay,
        overlayProps = {},
        children,
        className,
        container,
        disableAutoFocus = false,
        disableOverlayClick = false,
        disableEnforceFocus = false,
        disableEscapeKeyDown = false,
        disablePortal = false,
        disableRestoreFocus = false,
        disableScrollLock = false,
        hideOverlay = false,
        keepMounted = false,
        manager = defaultManager,
        onOverlayClick,
        onOpen,
        onClose,
        onEscapeKeyDown,
        onRendered,
        open,
        ...props
    },
    ref,
) {
    // Modal refs
    const modal = React.useRef({} as ModalInnerRef)
    const mountNodeRef = React.useRef(null)
    const modalRef = React.useRef(null)
    const handleRef = useForkRef(modalRef, ref)

    /**
     * Gets nearest DOM document that owns modal ref
     */
    const getDoc = () => ownerDocument(mountNodeRef.current)
    /**
     * Gets modal from ref
     */
    const getModal = () => {
        modal.current.modalRef = modalRef.current
        modal.current.mountNode = mountNodeRef.current

        return modal.current
    }

    /**
     * Calls the modal manager's mount logic to handle mounting to DOM
     */
    const handleMounted = () => {
        manager.mount(getModal(), { disableScrollLock })

        // Fix a bug on Chrome where the scroll isn't initially 0
        modalRef.current.scrollTop = 0
    }

    /**
     * Handles modal opening
     */
    const handleOpen = useEventCallback(() => {
        // Gets current container from DOM document structure
        const resolvedContainer = getContainer(container) || getDoc().body

        // Adds modal to container DOM
        manager.add(getModal(), resolvedContainer)

        // The element was already mounted
        if (modalRef.current) {
            handleMounted()
        }

        // Call `onOpen` trigger if defined
        if (typeof onOpen === 'function') {
            onOpen()
        }
    })

    /**
     * Determines whether or not modal is top modal (in case multiple modals are mounted)
     */
    const isTopModal = React.useCallback(() => manager.isTopModal(getModal()), [manager])

    /**
     * Handles setting portal reference
     */
    const handlePortalRef = useEventCallback((node) => {
        // Set current ref to portal
        mountNodeRef.current = node

        // Don't call rendered trigger functions
        if (!node) {
            return
        }

        // Call `onRendered` prop if defined
        if (onRendered) {
            onRendered()
        }

        // Handle mount if modal is top and open
        if (open && isTopModal()) {
            handleMounted()
        } else {
            // Hide below or unopened modal
            ariaHidden(modalRef.current, true)
        }
    })

    /**
     * Handles modal close logic
     */
    const handleClose = React.useCallback(() => {
        // Remove modal via manager
        manager.remove(getModal())

        // Call `onClose` trigger function if defined
        if (typeof onClose === 'function') {
            onClose()
        }
    }, [manager, onClose])

    // Handle close when React unmounts component
    React.useEffect(() => {
        return () => {
            handleClose()
        }
    }, [handleClose])

    // Rerenders component when `open` prop changes
    React.useEffect(() => {
        if (open) {
            handleOpen()
        } else {
            handleClose()
        }
    }, [open, handleClose, handleOpen])

    /**
     * Handles event when overlay is clicked
     *
     * @param {Event} event - DOM Event
     */
    const handleOverlayClick = (event) => {
        // Ignore when child of overlay is clicked
        if (event.target !== event.currentTarget) {
            return
        }

        // Trigger `onOverlayClick` if defined
        if (onOverlayClick) {
            onOverlayClick(event)
        }

        // Trigger `onClose` if defined and not disabled
        if (!disableOverlayClick && onClose) {
            onClose(event, 'overlayClick')
        }
    }

    /**
     * Handles key down event, for escape key
     *
     * @param {Event} event - DOM Event
     */
    const handleKeyDown = (event) => {
        // The handler doesn't take event.defaultPrevented into account:
        //
        // event.preventDefault() is meant to stop default behaviours like
        // clicking a checkbox to check it, hitting a button to submit a form,
        // and hitting left arrow to move the cursor in a text input etc.
        // Only special HTML elements have these default behaviors.
        if (event.key !== 'Escape' || !isTopModal()) {
            return
        }

        // Swallow the event, in case someone is listening for the escape key on the body.
        event.stopPropagation()

        // Trigger `onEscapeKeyDown` if defined
        if (onEscapeKeyDown) {
            onEscapeKeyDown(event)
        }

        // Trigger `onClose` if defined and not disabled
        if (!disableEscapeKeyDown && onClose) {
            onClose(event, 'escapeKeyDown')
        }
    }

    // Don't render component if closed and not keeping modal mounted
    if (!keepMounted && !open) {
        return null
    }

    // Props to set child with
    const childProps: ModalHandlerChildProps = {}

    if (children.props.tabIndex === undefined) {
        // Forces child to tab focus
        childProps.tabIndex = children.props.tabIndex || '-1'
    }

    /**
     * Layout:
     *  <Portal>
     *      <div modal-handler>
     *          <Overlay />
     *          <TrapFocus>
     *              ...
     *          </TrapFocus>
     *      </div>
     *  </Portal>
     */
    return (
        <Portal ref={handlePortalRef} container={container} disablePortal={disablePortal}>
            <ModalHandlerWrapper
                className={classNames(
                    'modal-handler',
                    {
                        'keep-mounted': keepMounted,
                    },
                    className,
                )}
                ref={handleRef}
                onKeyDown={handleKeyDown}
                role="presentation"
                {...props}
            >
                {hideOverlay ? null : (
                    <OverlayComponent open={open} onClick={handleOverlayClick} {...overlayProps} />
                )}
                <TrapFocus
                    disableEnforceFocus={disableEnforceFocus}
                    disableAutoFocus={disableAutoFocus}
                    disableRestoreFocus={disableRestoreFocus}
                    getDoc={getDoc}
                    isEnabled={isTopModal}
                    open={open}
                >
                    {React.cloneElement(children, {
                        ...childProps,
                    })}
                </TrapFocus>
            </ModalHandlerWrapper>
        </Portal>
    )
})

type ModalHandlerRef = HTMLDivElement

interface ModalHandlerProps extends BackyardBaseProps<ModalHandlerRef> {
    /**
     * Overlay component to display behind modal
     */
    OverlayComponent?: React.ElementType
    /**
     * Props to extend overlay component with
     * Useful if the `style` prop needs to be different
     */
    overlayProps?: OverlayProps
    /**
     * Children
     */
    children?: React.ReactElement
    /**
     * Container object or function of modal to display
     */
    container?: React.ReactInstance
    /**
     * Whether or not browser will auto focus on modal
     */
    disableAutoFocus?: boolean
    /**
     * Whether or not overlay click listener is active
     */
    disableOverlayClick?: boolean
    /**
     * Whether or not modal enforces focus (traps)
     */
    disableEnforceFocus?: boolean
    /**
     * Whether or not escape key listener is active
     */
    disableEscapeKeyDown?: boolean
    /**
     * Whether or not portal is used to bring modal to root DOM
     */
    disablePortal?: boolean
    /**
     * Whether or not modal restores focus
     */
    disableRestoreFocus?: boolean
    /**
     * Whether or not scroll lock is disabled
     */
    disableScrollLock?: boolean
    /**
     * Whether or not overlay is hidden
     */
    hideOverlay?: boolean
    /**
     * Whether or not modal continues to be mounted when closed
     * Useful for certain SEO conditions
     */
    keepMounted?: boolean
    /**
     * Class that manages modal state and properties
     */
    manager?: ModalManager
    /**
     * `onOverlayClick` trigger function, when overlay is clicked
     * @argument {Event} event - DOM event
     */
    onOverlayClick?: (event: React.MouseEvent) => void
    /**
     * `onClose` trigger function, when modal is closed
     * @argument {Event} event - DOM event
     * @argument {string} name - Name of event fired that caused modal to close
     */
    onClose?: (event?: ModalHandlerEvent, name?: string) => void
    /**
     * `onOpen` trigger function, when modal is closed
     */
    onOpen?: () => void
    /**
     * `onEscapeKeyDown` trigger function, when escape key is pressed while modal is active
     * @argument {Event} event - DOM event
     */
    onEscapeKeyDown?: (event: React.KeyboardEvent) => void
    /**
     * `onRendered` trigger function, when modal is closed
     */
    onRendered?: () => void
    /**
     * Whether or not modal is open
     */
    open: boolean
}

type ModalHandlerEvent = React.MouseEvent | React.KeyboardEvent

type ModalHandlerChildProps = {
    tabIndex?: string
}

type ModalInnerRef = {
    modalRef?: ModalHandlerRef
    mountNode?: PortalRef
}

ModalHandler.bdsName = 'ModalHandler'

export { ModalHandler }

export type { ModalHandlerProps, ModalHandlerRef }

export default ModalHandler
