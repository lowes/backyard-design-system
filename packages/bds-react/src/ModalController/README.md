# @lowes-tech/bds-react/ModalController

## Interfaces

- [ModalControllerContextValues](interfaces/ModalControllerContextValues.md)
- [ModalControllerProps](interfaces/ModalControllerProps.md)
- [ModalControllerProviderProps](interfaces/ModalControllerProviderProps.md)
- [ModalHandlerProps](interfaces/ModalHandlerProps.md)

## References

### default

Renames and re-exports [ModalController](README.md#modalcontroller)

## Type aliases

### ModalControllerRef

Ƭ **ModalControllerRef**: [`ModalHandlerRef`](README.md#modalhandlerref)

___

### ModalHandlerRef

Ƭ **ModalHandlerRef**: `HTMLDivElement`

## Variables

### ModalController

• **ModalController**: `BDSForwardRef`<[`ModalControllerProps`](interfaces/ModalControllerProps.md)\>

Backyard React Modal Controller

Extends from `ModalHandler` to offer and cleaner API to work with

Controls modal defined in `modal` prop with basic closing/opening handling

For more API controls integrated with this HOC, see `ModalHandler`

Example:
 <ModalController
     modal={(
         <Modal>
             <ModalHeader ... />
             <ModalBody ... />
             <ModalFooter ... />
         </Modal>
     )}
 />

___

### ModalControllerConsumer

• **ModalControllerConsumer**: `Consumer`<[`ModalControllerContextValues`](interfaces/ModalControllerContextValues.md)\> = `ModalControllerContext.Consumer`

___

### ModalControllerContext

• **ModalControllerContext**: `Context`<[`ModalControllerContextValues`](interfaces/ModalControllerContextValues.md)\>

___

### ModalHandler

• **ModalHandler**: `BDSForwardRef`<[`ModalHandlerProps`](interfaces/ModalHandlerProps.md)\>

Backyard React Modal Handler

This component is used to extend the higher ordered component `ModalController`
Abstracts some of the complexity for future extending, if needed

Heavily influenced by Material UI's Modal Component

**`see`** https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Modal/Modal.js

## Functions

### ModalControllerProvider

▸ `Const` **ModalControllerProvider**(`props`): `ReactElement`<[`ModalControllerProviderProps`](interfaces/ModalControllerProviderProps.md), `string` \| `JSXElementConstructor`<`any`\>\>

`ModalControllerProvider` Backyard React

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `props` | [`ModalControllerProviderProps`](interfaces/ModalControllerProviderProps.md) | Modal props |

#### Returns

`ReactElement`<[`ModalControllerProviderProps`](interfaces/ModalControllerProviderProps.md), `string` \| `JSXElementConstructor`<`any`\>\>

___

### useModalController

▸ `Const` **useModalController**(): [`ModalControllerContextValues`](interfaces/ModalControllerContextValues.md)

#### Returns

[`ModalControllerContextValues`](interfaces/ModalControllerContextValues.md)
