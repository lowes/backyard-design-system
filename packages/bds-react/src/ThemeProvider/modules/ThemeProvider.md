# Namespace: ThemeProvider

## Variables

### bdsName

â€¢ **bdsName**: `string`
