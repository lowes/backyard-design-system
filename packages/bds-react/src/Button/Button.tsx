import * as React from 'react'
import classnames from 'classnames'

import persistSyntheticEvent from '../utils/functions/persistSyntheticEvent'
import type { BackyardBaseProps } from '../utils/typings/BackyardProps'
import type { BDSForwardRef } from '../utils/typings/BDSComponentProps'
import { useBackyardTheme, getShape } from '../ThemeProvider'

import ButtonBase from './styles/ButtonBase'
import { useForwardedRef } from '../utils/hooks'

/**
 * Backyard React Button
 *
 * Button for interacting via text.
 *
 * For Icon only buttons, see `IconButton` instead.
 *
 *  <Button
 *      variant='primary'
 *      color='commerce'
 *      iconAfter={<ChevronRight />}
 *  >
 *      Checkout
 *  </Button>
 */
const Button: BDSForwardRef<ButtonProps> = React.forwardRef<ButtonRef, ButtonProps>(function Button(
    {
        children,
        className,
        variant = 'primary',
        iconAfter: iconAfterProp,
        iconBefore: iconBeforeProp,
        shape: shapeProp, // = 'rounded',
        color = 'interactive',
        disabled = false,
        size = 'medium',
        fullWidth = false,
        elevation = false,
        style,
        type = 'button',
        onClick,
        onKeyDown,
        onFocus,
        ...props
    },
    ref,
) {
    // Get Backyard Theme Context
    const theme = useBackyardTheme()
    const { shape: shapeContext } = theme.context

    const innerRef = useForwardedRef(ref)

    // Calculate shape
    const shape = getShape(shapeProp, shapeContext) as ButtonProps['shape']

    // If `iconBefore` defined, insert into <span>
    const iconBefore = iconBeforeProp && <span className='btn-start-icon'>{iconBeforeProp}</span>
    // If `iconAfter` defined, insert into <span>
    const iconAfter = iconAfterProp && <span className='btn-end-icon'>{iconAfterProp}</span>

    /**
     * Handles `onFocus` event of button
     * Triggers defined `onFocus` from props if available
     *
     * @param {Event} event - DOM event
     */
    const handleFocus = (event) => {
        // Persist event propagation
        persistSyntheticEvent(event)

        // If not disabled,
        if (!disabled) {
            // If `onFocus` prop defined,
            if (onFocus) {
                // Trigger `onFocus` prop function
                onFocus(event)
            }
        }
    }

    /**
     * Handles `onClick` event of button
     * Triggers defined `onClick` from props if available
     *
     * @param {Event} event - DOM event
     */
    const handleClick = (event) => {
        // Persist event propagation
        persistSyntheticEvent(event)

        // If not disabled,
        if (!disabled) {
            // If `onClick` prop defined,
            if (onClick) {
                // Trigger `onClick` prop function
                onClick(event)
            }
        }
    }

    /**
     * Handles `onKeyDown` event of button
     * Triggers defined `onKeyDown` from props if available
     *
     * @param {Event} event - DOM event
     */
    const handleKeyDown = (event) => {
        // Persist event propagation
        persistSyntheticEvent(event)

        if (!disabled) {
            // If `onKeyDown` prop defined,
            if (onKeyDown) {
                // Trigger `onKeyDown` prop function
                onKeyDown(event)
            }
        }
    }

    const createVariantStyles = (variant: string, color: string) => {
        switch(variant) {
            case 'secondary':
            case 'tertiary':
            case 'ghost':
            case 'inverse':
                return {
                    '--btn-base': (color === 'neutral') ? theme.color.border_default : theme.color[`action_${color}`],
                    '--btn-base-subdued': (variant === 'inverse') ? theme.color.surface_default : theme.color[`action_${color}_subdued`],
                    '--btn-hover': theme.color[`action_${color}_subdued_hover`],
                    '--btn-pressed': theme.color[`action_${color}_subdued_pressed`],
                    '--btn-focus': (color === 'neutral') ? theme.color.border_default : theme.color[`action_${color}`],
                    '--btn-text': (color === 'neutral') ? theme.color.text_primary : theme.color[`text_${color}`],
                    '--btn-icon': (color === 'neutral') ? theme.color.icon_primary : theme.color[`icon_${color}`],
                    '--btn-disabled': theme.color.text_disabled,
                    '--btn-disabled-subdued': theme.color.surface_subdued
                }

            default:
                return {
                    '--btn-base': theme.color[`action_${color}`],
                    '--btn-hover': theme.color[`action_${color}_hover`],
                    '--btn-pressed': theme.color[`action_${color}_pressed`],
                    '--btn-focus': (color === 'neutral') ? theme.color.border_default : theme.color.focus_inverse,
                    '--btn-text': (color === 'neutral') ? theme.color.text_primary : theme.color.text_solid_inverse,
                    '--btn-icon': (color === 'neutral') ? theme.color.icon_primary : theme.color.icon_solid_inverse,
                    '--btn-disabled': theme.color.text_disabled,
                    '--btn-onDisabled': theme.color.text_disabled_inverse
                }
        }
    }

    // Side effect for when button changes disabled state,
    React.useEffect(() => {
        // If button is disabled,
        if (disabled) {
            // Blur button from focus
            innerRef.current?.blur()
        }
    }, [disabled])

    /**
     * Layout:
     *
     *  <ButtonBase>
     *      <span>
     *          <icon (start) />
     *          ...
     *          <icon (end) />
     *      </span>
     *  </ButtonBase>
     */
    return (
        <ButtonBase
            className={classnames(
                'backyard',
                'button',
                `size--${size}`,
                `variant--${variant}`,
                `color--${color}`,
                `shape--${shape}`,
                fullWidth ? 'full-width' : null,
                (elevation && (variant === 'primary' || variant === 'tertiary' || variant === 'inverse')) ?
                    'elevation' : '',
                className,
                {
                    disabled,
                },
            )}
            variant={variant}
            disabled={disabled}
            color={color}
            size={size}
            ref={innerRef}
            onClick={handleClick}
            onKeyDown={handleKeyDown}
            onFocus={handleFocus}
            type={type}
            style={{
                ...createVariantStyles(variant, color),
                ...style,
            }}
            {...props}
        >
            <span className='label btn-label'>
                {iconBefore}
                {children}
                {iconAfter}
            </span>
        </ButtonBase>
    )
})

type ButtonRef = HTMLButtonElement

type ButtonOverrideProps = 'size' | 'color'

interface ButtonProps extends BackyardBaseProps<ButtonRef, ButtonOverrideProps> {
    /**
     * React Children
     */
    children?: React.ReactNode
    /**
     * DOM Class Name
     */
    className?: string
    /**
     * Variant of button
     * - Primary -> (Default) Solid Button w/ Background w/o Borders
     * - Secondary -> Outlined Button w/o Background w/ Borders
     * - Ghost -> No Background, No Borders
     */
    variant?: 'primary' | 'secondary' | 'tertiary' | 'ghost' | 'inverse'
    /**
     * Icon Node to display on right-side of button text
     *
     * Note: Do not use for an icon-only button, use `IconButton` for that
     */
    iconAfter?: React.ReactNode
    /**
     * Icon Node to display on left-side of button text
     *
     * Note: Do not use for an icon-only button, use `IconButton` for that
     */
    iconBefore?: React.ReactNode
    /**
     * Type of button
     * - Interactive -> (Default) Lowe's Primary Color Palette
     * - Commerce -> Lowe's Commerce Color Palette
     * - Contrast -> High Contrast Black/White Color Palette
     */
    color?: 'interactive' | 'green' | 'red' | 'neutral'
    /**
     * Shape of button
     * - Rounded -> Medium Border Radius
     * - Squared -> (Default) Zero Border Radius
     * - Circle -> Maximum Border Radius (FAB Button)
     */
    shape?: 'rounded' | 'squared' | 'circle' | 'custom'
    /**
     * Whether or not button is disabled
     */
    disabled?: boolean
    /**
     * Size of Button
     * - Small -> Small Height/Spacing Button
     * - Medium -> (Default) Medium Height/Spacing Button
     * - Large -> Large Height/Spacing Button
     */
    size?: 'small' | 'medium' | 'large' | 'jumbo' | 'full-width' | 'extra_small'
    /**
     *
     */
    type?: 'button' | 'submit' | 'reset'
    /**
     * Makes button 100% the width of the container
     */
    fullWidth?: boolean
    /**
     * Adds elevation to the button
     */
    elevation?: boolean
    /**
     * `onClick` trigger function
     * @argument {Event} event - DOM event
     */
    onClick?: (event: React.MouseEvent) => void
    /**
     * `onKeyDown` trigger function
     * @argument {Event} event - DOM event
     */
    onKeyDown?: (event: React.KeyboardEvent) => void
    /**
     * `onFocus` trigger function
     * @argument {Event} event - DOM event
     */
    onFocus?: (event: React.FocusEvent) => void
}

Button.bdsName = 'Button'

export { Button }

export type { ButtonProps, ButtonRef }

export default Button
