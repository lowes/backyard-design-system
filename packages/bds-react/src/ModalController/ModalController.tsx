import * as React from 'react'
import classNames from 'classnames'

import { useBackyardTheme } from '../ThemeProvider'
import type { BDSForwardRef } from '../utils/typings/BDSComponentProps'

import ModalHandler, { ModalHandlerRef, ModalHandlerProps } from './ModalHandler'
import ModalControllerProvider from './ModalControllerProvider'

/**
 * Backyard React Modal Controller
 *
 * Extends from `ModalHandler` to offer and cleaner API to work with
 *
 * Controls modal defined in `modal` prop with basic closing/opening handling
 *
 * For more API controls integrated with this HOC, see `ModalHandler`
 *
 * Example:
 *  <ModalController
 *      modal={(
 *          <Modal>
 *              <ModalHeader ... />
 *              <ModalBody ... />
 *              <ModalFooter ... />
 *          </Modal>
 *      )}
 *  />
 */
const ModalController: BDSForwardRef<ModalControllerProps> = React.forwardRef<
    ModalControllerRef,
    ModalControllerProps
>(function ModalController(
    {
        className,
        modal: modalProp,
        open: openProp = false,
        onClose,
        onOpen,
        style: styleProp,
        modalStyle: modalStyleProp,
        position = {
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
        },
        context,
        override,
        ...props
    },
    ref,
) {
    // Get Backyard Theme
    const theme = useBackyardTheme({ validate: true, name: 'ModalController' })

    // Track whether or not modal is open
    const [open, setOpen] = React.useState(openProp)

    /**
     * Close the modal
     */
    const closeModal = () => {
        setOpen(false)

        if (onClose) {
            onClose()
        }
    }

    /**
     * Open the modal
     */
    const openModal = () => {
        setOpen(true)

        if (onOpen) {
            onOpen()
        }
    }

    // Side effect for handling whether to open or close modal when `openProp` changes from external sources
    React.useEffect(() => (openProp ? openModal() : closeModal()), [openProp])

    // Memoize style prop
    const style = React.useMemo<React.CSSProperties>(
        () => ({
            position: 'fixed',
            top: 0,
            bottom: 0,
            left: 0,
            right: 0,
            zIndex: theme.zIndex.modal,
            ...styleProp,
        }),
        [theme.name, styleProp],
    )

    // Memoize modal style
    const modalStyle = React.useMemo(
        () => ({
            ...position,
            zIndex: theme.zIndex.modal + 1,
            ...modalProp?.props?.style,
            ...modalStyleProp,
        }),
        [theme.zIndex, modalStyleProp, position],
    )

    // Memoize modal render
    const modal = React.useMemo(
        () =>
            React.cloneElement(modalProp, {
                style: modalStyle,
                onCloseClick: () => closeModal(),
            }),
        [modalStyle],
    )

    /**
     * Layout:
     *  <ModalHandler>
     *      ...
     *  </ModalHandler>
     */
    return (
        <ModalControllerProvider open={open} context={context} override={override}>
            <ModalHandler
                className={classNames(
                    'modal-controller',
                    {
                        opened: open,
                        closed: !open,
                    },
                    className,
                )}
                open={open}
                onEscapeKeyDown={() => closeModal()}
                onOverlayClick={() => closeModal()}
                style={style}
                {...props}
                ref={ref}
            >
                {modal}
            </ModalHandler>
        </ModalControllerProvider>
    )
})

type ModalControllerRef = ModalHandlerRef

type ModalControllerOverrideProps = 'open'

interface ModalControllerProps extends Omit<ModalHandlerProps, ModalControllerOverrideProps> {
    /**
     * DOM Class Name
     */
    className?: string
    /**
     * Modal component to control
     */
    modal: React.ReactElement
    /**
     * Whether or not modal is open
     */
    open?: boolean
    /**
     * Style of modal controller
     */
    style?: React.CSSProperties
    /**
     * Style of the modal
     */
    modalStyle?: React.CSSProperties
    /**
     * Position of modal on page
     *
     * @default 'center of page'
     */
    position?: React.CSSProperties
    /**
     * Props to provide context for
     */
    context?: Record<string, any>
    /**
     * Props to override children with
     */
    override?: Record<string, any>
}

ModalController.bdsName = 'ModalController'

export { ModalController, ModalControllerProps, ModalControllerRef }

export default ModalController
