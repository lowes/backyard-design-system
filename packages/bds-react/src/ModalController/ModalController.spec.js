import * as React from 'react'
import { Settings } from '@lowes-tech/bds-icons'

import { render } from '../../test-utils'
import ThemeProvider from '../ThemeProvider'
import Button from '../Button'
import {
    Modal,
    ModalBody,
    ModalHeader,
    ModalFooter
 } from '../Modal'

import ModalController from './ModalController'

describe('ModalController Snapshots', () => {
    test('ModalController default snapshot', () => {
        const { asFragment } = render(
            <ModalController
                open
                modal={(
                    <Modal>
                        <ModalHeader
                            label="Label"
                            icon={<Settings />}
                        >
                            Modal Heading
                        </ModalHeader>
                        <ModalBody>
                            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse cursus fermentum risus,
                            sit amet fringilla nunc pellentesque quis. Duis quis odio ultrices, cursus lacus ac, posuere felis.
                            Donec dignissim libero in augue mattis, a molestie metus vestibulum. Aliquam placerat felis
                            ultrices lorem condimentum, nec ullamcorper felis porttitor.
                        </ModalBody>
                        <ModalFooter>
                            <Button
                                variant="secondary"
                                color="contrast"
                            >
                                Cancel
                            </Button>
                            <Button>Action</Button>
                        </ModalFooter>
                    </Modal>
                )}
            />
        )

        expect(asFragment()).toMatchSnapshot()
    })

    describe('themes', () => {
        const themes = ['light', 'dark']

        themes.forEach((theme) => {
            test(`ModalController ${theme} theme snapshot`, () => {
                const { asFragment } = render(
                    <ThemeProvider theme={theme}>
                        <React.Fragment>
                            <ModalController
                                open
                                modal={(
                                    <Modal>
                                        <ModalHeader
                                            label="Label"
                                            icon={<Settings />}
                                        >
                                            Modal Heading
                                        </ModalHeader>
                                        <ModalBody>
                                            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse cursus fermentum risus,
                                            sit amet fringilla nunc pellentesque quis. Duis quis odio ultrices, cursus lacus ac, posuere felis.
                                            Donec dignissim libero in augue mattis, a molestie metus vestibulum. Aliquam placerat felis
                                            ultrices lorem condimentum, nec ullamcorper felis porttitor.
                                        </ModalBody>
                                        <ModalFooter>
                                            <Button
                                                variant="secondary"
                                                color="contrast"
                                            >
                                                Cancel
                                            </Button>
                                            <Button>Action</Button>
                                        </ModalFooter>
                                    </Modal>
                                )}
                            />
                        </React.Fragment>
                    </ThemeProvider>
                )

                expect(asFragment()).toMatchSnapshot()
            })
        })
    })

    describe('extra props', () => {
        test('ModalController DOM props (id, className) snapshot', () => {
            const { asFragment } = render(
                <ModalController
                    open
                    className="modal-controller-test-class"
                    modal={(
                        <Modal>
                            <ModalHeader
                                label="Label"
                                icon={<Settings />}
                            >
                                Modal Heading
                            </ModalHeader>
                            <ModalBody>
                                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse cursus fermentum risus,
                                sit amet fringilla nunc pellentesque quis. Duis quis odio ultrices, cursus lacus ac, posuere felis.
                                Donec dignissim libero in augue mattis, a molestie metus vestibulum. Aliquam placerat felis
                                ultrices lorem condimentum, nec ullamcorper felis porttitor.
                            </ModalBody>
                            <ModalFooter>
                                <Button
                                    variant="secondary"
                                    color="contrast"
                                >
                                    Cancel
                                </Button>
                                <Button>Action</Button>
                            </ModalFooter>
                        </Modal>
                    )}
                />
            )

            expect(asFragment()).toMatchSnapshot()
        })
    })
})
