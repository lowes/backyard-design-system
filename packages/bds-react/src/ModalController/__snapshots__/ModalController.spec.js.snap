// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalController Snapshots ModalController default snapshot 1`] = `<DocumentFragment />`;

exports[`ModalController Snapshots extra props ModalController DOM props (id, className) snapshot 1`] = `<DocumentFragment />`;

exports[`ModalController Snapshots themes ModalController dark theme snapshot 1`] = `<DocumentFragment />`;

exports[`ModalController Snapshots themes ModalController light theme snapshot 1`] = `<DocumentFragment />`;
